fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,281 @@
+                    _ => panic!("unexpected message type"),
+                };
+
+                // execute the message
+                let (res, _) = state.execute_implicit(message).unwrap();
+                assert!(
+                    res.msg_receipt.exit_code.is_success(),
+                    "{:?}",
+                    res.failure_info
+                );
+
+                // parse the message receipt data and make sure the contract was deployed to the expected address
+                let res = fvm_ipld_encoding::from_slice::<eam::CreateReturn>(
+                    &res.msg_receipt.return_data,
+                )
+                .unwrap();
+                assert_eq!(
+                    res.delegated_address(),
+                    Address::from_str(CONTRACT_ADDRESS).unwrap()
+                );
+
+                // lets upgrade the app version as well
+                let app_version = 1;
+
+                Ok(app_version)
+            })
+            .unwrap(),
+        )
+        .unwrap();
+
+    upgrade_scheduler
+        .add(
+            Upgrade::new("mychain".to_string(), 2, |state| {
+                println!(
+                    "[Upgrade at height {}] Sends a balance",
+                    state.block_height()
+                );
+
+                // build the calldata for the send_coin function
+                let (client, _mock) = ethers::providers::Provider::mocked();
+                let simple_coin = SimpleCoin::new(EthAddress::from_id(101), client.into());
+                let call = simple_coin.send_coin(
+                    // the address we are sending the balance to (which is us in this case)
+                    eth_addr_to_h160(&EthAddress::from(my_secret_key().public_key())),
+                    // the amount we are sending
+                    U256::from(SEND_BALANCE_AMOUNT),
+                );
+
+                // create a message for sending the balance
+                let mut mf = MessageFactory::new_secp256k1(my_secret_key(), 1, state.chain_id());
+                let message = match mf
+                    .fevm_invoke(
+                        Address::from_str(CONTRACT_ADDRESS).unwrap(),
+                        call.calldata().unwrap().0,
+                        TokenAmount::default(),
+                        GAS_PARAMS.clone(),
+                    )
+                    .unwrap()
+                {
+                    ChainMessage::Signed(signed) => signed.into_message(),
+                    _ => panic!("unexpected message type"),
+                };
+
+                // execute the message
+                let (res, _) = state.execute_implicit(message).unwrap();
+                assert!(
+                    res.msg_receipt.exit_code.is_success(),
+                    "{:?}",
+                    res.failure_info
+                );
+
+                // this upgrade didn't affect our app version, so return 0
+                Ok(0)
```

Comment:
Right. Maybe return `None` - to make it more difficult for people to misunderstand. In my experience people tend to forget about 0 being a special case. I used this for the Solidity contracts when we have no change in the validator set, because Solidity doesn't have `Option`, but had to explain it many times later.